### PR TITLE
[WIP]Introducing a separation of concerns between data sources that are gi…

### PR DESCRIPTION
**Pull request by RogerDunn**
…ven a dataframe's schema upon write, from FileDataSourceV2, in part to allow data sources that require this notion to be developed in Java (FileDataSourceV2 is not compatible with Java because of its private Table member)

### CODE REVIEW

**Comment by RogerDunn** [comment]
@AmplabJenkins , thank you for the speedy  note on this patch.  Please, there is no current urgency on this patch.  I marked it as a [WIP] PR because I'm currently in the process of reviewing this change (weighed against other alternatives) with another Spark developer.  I'll keep the notes up to date on this PR.  Thank you!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/SchemaOnWriteDataSource.scala [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ */
+package org.apache.spark.sql.execution.datasources.v2
+
+import org.apache.spark.sql.connector.catalog.TableProvider
+import org.apache.spark.sql.sources.DataSourceRegister
+
+/**
+ * A data source that is given the DataFrame's schema upon write operations.
+ */
+trait SchemaOnWriteDataSource extends TableProvider with DataSourceRegister {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/SchemaOnWriteDataSource.scala:25]: One option is to add this trait, and another option is to officially make `FileDataSourceV2` a public developer API. Does this "schema-on-write" behavior make sense for non-file data sources?

cc @gengliangwang

Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/SchemaOnWriteDataSource.scala:25]: +1, making `FileDataSourceV2` a public developer API seems simpler

**Comment by cloud-fan** [comment]
The schema inference is already in the DS v2 API `TableProvider` which is not bound to file source. I think this feature "skip schema inference on write" should not bound to file source as well.

Instead of adding a marker trait, I think we can just reuse `TableProvider.supportsExternalMetadata`, which indicates that the schema inference is expensive. If `supportsExternalMetadata == true`, the `DataFrameWriter.save` should not infer schema and just pass the input schema to `TableProvider`.

**Comment by RogerDunn** [comment]
@cloud-fan Your idea sounds just right.  Are you proposing to make that change in the work you're already doing (in which case I'll remove this PR)?

**Comment by cloud-fan** [comment]
@gengliangwang is creating a PR to implement this idea.

**Comment by gengliangwang** [comment]
I have created https://github.com/apache/spark/pull/30273 for this.

**Comment by RogerDunn** [comment]
@gengliangwang your PR is better, thank you!

**Pull request closed without merging**